staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go

```diff
@@ -772,7 +753,106 @@ func (vs *VSphere) InstanceShutdownByProviderID(ctx context.Context, providerID
 
 // InstanceMetadataByProviderID returns metadata of the specified instance.
 func (vs *VSphere) InstanceMetadataByProviderID(ctx context.Context, providerID string) (*cloudprovider.InstanceMetadata, error) {
-	return nil, fmt.Errorf("unimplemented")
+	if providerID == "" {
+		return nil, fmt.Errorf("couldn't compute InstanceMetadata for empty providerID")
+	}
+
+	nodeName, err := vs.GetNodeNameFromProviderID(providerID)
+	if err != nil {
+		klog.Errorf("Error while getting nodename by providerID %s", providerID)
+		return nil, err
+	}
+	md, err := vs.instanceMetadataByNodeName(ctx, nodeName)
+	if vclib.IsManagedObjectNotFoundError(err) {
+		if err = vs.nodeManager.RediscoverNode(k8stypes.NodeName(nodeName)); err != nil {
+			return nil, err
+		}
+		return vs.instanceMetadataByNodeName(ctx, nodeName)
+	}
+	return md, err
+}
+
+func (vs *VSphere) instanceMetadataByNodeName(ctx context.Context, nodeName string) (*cloudprovider.InstanceMetadata, error) {
```

Comment:
I'm not familiar with this code, but it seems you can probably share a bit more than the nodeAddressesFromVM part

Reply:
You're right, all logic of `NodeAddresses` can be reused, cleaner now I think.

Reply:
Thanks!